From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Otto=20Pfl=C3=BCger?= <otto.pflueger@abscue.de>
Date: Tue, 15 Apr 2025 10:52:16 +0200
Subject: [PATCH 079/225] ASoC: sprd: add UMS9230 machine driver
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Add a machine driver for the UMS9230 SoC. Only include the VBC DAIs by
default since these are always present on the SoC. The device-specific
codec links are read from the device tree.

Signed-off-by: Otto Pflüger <otto.pflueger@abscue.de>
---
 sound/soc/sprd/Kconfig            |  29 +++--
 sound/soc/sprd/Makefile           |   8 +-
 sound/soc/sprd/sprd-card-common.c | 166 +++++++++++++++++++++++++++++
 sound/soc/sprd/sprd-card-common.h |  13 +++
 sound/soc/sprd/ums9230-card.c     | 169 ++++++++++++++++++++++++++++++
 5 files changed, 376 insertions(+), 9 deletions(-)
 create mode 100644 sound/soc/sprd/sprd-card-common.c
 create mode 100644 sound/soc/sprd/sprd-card-common.h
 create mode 100644 sound/soc/sprd/ums9230-card.c

diff --git a/sound/soc/sprd/Kconfig b/sound/soc/sprd/Kconfig
index 21f5b195d..760b49906 100644
--- a/sound/soc/sprd/Kconfig
+++ b/sound/soc/sprd/Kconfig
@@ -1,14 +1,20 @@
 # SPDX-License-Identifier: GPL-2.0-only
-menu "Spreadtrum"
-
-config SND_SOC_SPRD
-	tristate "SoC Audio for the Spreadtrum SoC chips"
+menuconfig SND_SOC_SPRD
+	tristate "Spreadtrum"
 	depends on ARCH_SPRD || COMPILE_TEST
-	select SND_SOC_COMPRESS
 	help
 	  Say Y or M if you want to add support for codecs attached to
 	  the Spreadtrum SoCs' Audio interfaces.
 
+if SND_SOC_SPRD
+
+config SND_SOC_SPRD_COMMON
+	tristate
+
+config SND_SOC_SPRD_PLATFORM
+	tristate
+	select SND_SOC_COMPRESS
+
 config SND_SOC_SPRD_MCDT
 	tristate "Spreadtrum multi-channel data transfer support"
 	depends on SND_SOC_SPRD
@@ -17,6 +23,17 @@ config SND_SOC_SPRD_MCDT
 	  is used for sound stream transmission between audio subsystem
 	  and other AP/CP subsystem.
 
+config SND_SOC_SPRD_UMS9230
+	tristate "SoC Audio support for UMS9230"
+	depends on SND_SOC_SPRD
+	select SND_SOC_SPRD_COMMON
+	select SND_SOC_SPRD_MCDT
+	select SND_SOC_SPRD_PLATFORM
+	select SND_SOC_SPRD_VBC_V4_DSP
+	help
+	  Say Y here to enable audio support for typical systems such as
+	  mobile phones based on the Unisoc UMS9230 SoCs
+
 config SND_SOC_SPRD_VBC_V4_DSP
 	tristate "Unisoc VBC v4 (DSP) support"
 	depends on SPRD_AGDSP
@@ -25,4 +42,4 @@ config SND_SOC_SPRD_VBC_V4_DSP
 	  Say Y here to add support for DSP-based audio in SoCs such as
 	  UMS9230.
 
-endmenu
+endif
diff --git a/sound/soc/sprd/Makefile b/sound/soc/sprd/Makefile
index 450895f62..1e32d919a 100644
--- a/sound/soc/sprd/Makefile
+++ b/sound/soc/sprd/Makefile
@@ -1,11 +1,13 @@
 # SPDX-License-Identifier: GPL-2.0
 # Spreadtrum Audio Support
 
+snd-soc-sprd-common-y := sprd-card-common.o
 snd-soc-sprd-platform-y := sprd-pcm-dma.o sprd-pcm-compress.o
 snd-soc-vbc-v4-dsp-y := vbc-v4-dsp.o
+snd-soc-ums9230-y := ums9230-card.o
 
-obj-$(CONFIG_SND_SOC_SPRD) += snd-soc-sprd-platform.o
-
+obj-$(CONFIG_SND_SOC_SPRD_COMMON) += snd-soc-sprd-common.o
 obj-$(CONFIG_SND_SOC_SPRD_MCDT) += sprd-mcdt.o
-
+obj-$(CONFIG_SND_SOC_SPRD_PLATFORM) += snd-soc-sprd-platform.o
+obj-$(CONFIG_SND_SOC_SPRD_UMS9230) += snd-soc-ums9230.o
 obj-$(CONFIG_SND_SOC_SPRD_VBC_V4_DSP) += snd-soc-vbc-v4-dsp.o
diff --git a/sound/soc/sprd/sprd-card-common.c b/sound/soc/sprd/sprd-card-common.c
new file mode 100644
index 000000000..23bed24dd
--- /dev/null
+++ b/sound/soc/sprd/sprd-card-common.c
@@ -0,0 +1,166 @@
+// SPDX-License-Identifier: GPL-2.0-only
+/*
+ * Copyright (C) 2025 Otto Pflüger
+ */
+
+#include <linux/module.h>
+#include <linux/of_reserved_mem.h>
+#include <sound/pcm_params.h>
+#include <sound/soc.h>
+
+#include "sprd-card-common.h"
+
+static int sprd_hw_params_fixup_24bit(struct snd_soc_pcm_runtime *rtd,
+				      struct snd_pcm_hw_params *params)
+{
+	struct snd_interval *rate = hw_param_interval(params,
+					SNDRV_PCM_HW_PARAM_RATE);
+	struct snd_interval *channels = hw_param_interval(params,
+					SNDRV_PCM_HW_PARAM_CHANNELS);
+	struct snd_mask *fmt = hw_param_mask(params,
+					SNDRV_PCM_HW_PARAM_FORMAT);
+
+	rate->min = rate->max = 48000;
+	channels->min = channels->max = 2;
+	snd_mask_none(fmt);
+	snd_mask_set(fmt, SNDRV_PCM_FORMAT_S24_LE);
+
+	return 0;
+}
+
+static int sprd_hw_params_fixup_16bit(struct snd_soc_pcm_runtime *rtd,
+				      struct snd_pcm_hw_params *params)
+{
+	struct snd_interval *rate = hw_param_interval(params,
+					SNDRV_PCM_HW_PARAM_RATE);
+	struct snd_interval *channels = hw_param_interval(params,
+					SNDRV_PCM_HW_PARAM_CHANNELS);
+	struct snd_mask *fmt = hw_param_mask(params,
+					SNDRV_PCM_HW_PARAM_FORMAT);
+
+	rate->min = rate->max = 48000;
+	channels->min = channels->max = 2;
+	snd_mask_none(fmt);
+	snd_mask_set(fmt, SNDRV_PCM_FORMAT_S16_LE);
+
+	return 0;
+}
+
+int sprd_card_parse_of(struct snd_soc_card *card)
+{
+	struct device_node *np;
+	struct device_node *codec = NULL;
+	struct device_node *platform = NULL;
+	struct device *dev = card->dev;
+	struct snd_soc_dai_link *link;
+	const char *link_name;
+	int ret, i;
+	u32 val;
+
+	ret = of_reserved_mem_device_init(card->dev);
+	if (ret)
+		dev_warn(card->dev,
+			 "no reserved DMA memory for sound card device\n");
+
+	ret = snd_soc_of_parse_card_name(card, "model");
+	if (ret) {
+		dev_err(dev, "error parsing card name: %d\n", ret);
+		return ret;
+	}
+
+	/* DAPM routes */
+	if (of_property_present(dev->of_node, "audio-routing")) {
+		ret = snd_soc_of_parse_audio_routing(card, "audio-routing");
+		if (ret)
+			return ret;
+	}
+
+	if (of_property_present(dev->of_node, "widgets")) {
+		ret = snd_soc_of_parse_audio_simple_widgets(card, "widgets");
+		if (ret)
+			return ret;
+	}
+
+	ret = snd_soc_of_parse_pin_switches(card, "pin-switches");
+	if (ret)
+		return ret;
+
+	ret = snd_soc_of_parse_aux_devs(card, "aux-devs");
+	if (ret)
+		return ret;
+
+	platform = of_parse_phandle(dev->of_node, "sprd,pcm-platform", 0);
+	if (!platform) {
+		dev_err(dev, "sprd,pcm-platform property missing or invalid\n");
+		return -EINVAL;
+	}
+
+	/* Enable platform PCM DMA for frontend DAIs that need it */
+	for_each_card_prelinks(card, i, link) {
+		if (link->no_pcm || link->c2c_params)
+			link->num_platforms = 0;
+		else
+			link->platforms->of_node = platform;
+	}
+
+	/* Codec links */
+	for_each_available_child_of_node(dev->of_node, np) {
+		ret = of_property_read_string(np, "link-name", &link_name);
+		if (ret) {
+			dev_err(card->dev, "error getting codec link name\n");
+			goto err_put_np;
+		}
+
+		for_each_card_prelinks(card, i, link) {
+			if (!strcmp(link_name, link->name))
+				break;
+		}
+
+		if (i >= card->num_links) {
+			dev_err(card->dev, "link %s does not exist\n", link_name);
+			ret = -EINVAL;
+			goto err_put_np;
+		}
+
+		ret = of_property_read_u32(np, "sprd,fixed-fmt", &val);
+		if (!ret) {
+			if (val == 24) {
+				link->be_hw_params_fixup =
+					sprd_hw_params_fixup_24bit;
+			} else if (val == 16) {
+				link->be_hw_params_fixup =
+					sprd_hw_params_fixup_16bit;
+			} else {
+				dev_err(card->dev, "%d-bit I2S not supported\n",
+					val);
+				ret = -EINVAL;
+				goto err_put_np;
+			}
+		}
+
+		link->dai_fmt = snd_soc_daifmt_parse_format(np, NULL);
+		link->dai_fmt |= SND_SOC_DAIFMT_CBP_CFP;
+
+		codec = of_get_child_by_name(np, "codec");
+		if (codec) {
+			ret = snd_soc_of_get_dai_link_codecs(dev, codec, link);
+			if (ret < 0) {
+				dev_err_probe(card->dev, ret,
+					      "%s: codec dai not found\n", link->name);
+				goto err_put_codec;
+			}
+			of_node_put(codec);
+		}
+	}
+
+	return 0;
+err_put_codec:
+	of_node_put(codec);
+err_put_np:
+	of_node_put(np);
+	return ret;
+}
+EXPORT_SYMBOL_GPL(sprd_card_parse_of);
+
+MODULE_DESCRIPTION("ASoC Unisoc card helper functions");
+MODULE_LICENSE("GPL");
diff --git a/sound/soc/sprd/sprd-card-common.h b/sound/soc/sprd/sprd-card-common.h
new file mode 100644
index 000000000..30a888218
--- /dev/null
+++ b/sound/soc/sprd/sprd-card-common.h
@@ -0,0 +1,13 @@
+// SPDX-License-Identifier: GPL-2.0-only
+/*
+ * Copyright (C) 2025 Otto Pflüger
+ */
+
+#ifndef __SPRD_CARD_COMMON_H
+#define __SPRD_CARD_COMMON_H
+
+struct snd_soc_card;
+
+int sprd_card_parse_of(struct snd_soc_card *card);
+
+#endif
diff --git a/sound/soc/sprd/ums9230-card.c b/sound/soc/sprd/ums9230-card.c
new file mode 100644
index 000000000..51fe6d692
--- /dev/null
+++ b/sound/soc/sprd/ums9230-card.c
@@ -0,0 +1,169 @@
+// SPDX-License-Identifier: GPL-2.0-only
+/*
+ * Copyright (C) 2025 Otto Pflüger
+ */
+
+#include <linux/module.h>
+#include <sound/soc.h>
+
+#include "sprd-card-common.h"
+
+SND_SOC_DAILINK_DEFS(fe_fast_playback,
+	DAILINK_COMP_ARRAY(COMP_CPU("VBC_FE_FAST_PLAYBACK")),
+	DAILINK_COMP_ARRAY(COMP_DUMMY()),
+	DAILINK_COMP_ARRAY(COMP_EMPTY()));
+
+SND_SOC_DAILINK_DEFS(fe_capture_dsp,
+	DAILINK_COMP_ARRAY(COMP_CPU("VBC_FE_CAPTURE_DSP")),
+	DAILINK_COMP_ARRAY(COMP_DUMMY()),
+	DAILINK_COMP_ARRAY(COMP_EMPTY()));
+
+SND_SOC_DAILINK_DEFS(voice_hostless,
+	DAILINK_COMP_ARRAY(COMP_CPU("VBC_FE_VOICE")),
+	DAILINK_COMP_ARRAY(COMP_DUMMY()),
+	DAILINK_COMP_ARRAY(COMP_EMPTY()));
+
+SND_SOC_DAILINK_DEFS(voice_capture,
+	DAILINK_COMP_ARRAY(COMP_CPU("VBC_FE_VOICE_CAPTURE")),
+	DAILINK_COMP_ARRAY(COMP_DUMMY()),
+	DAILINK_COMP_ARRAY(COMP_EMPTY()));
+
+SND_SOC_DAILINK_DEFS(be_iis0,
+	DAILINK_COMP_ARRAY(COMP_CPU("VBC_IIS0")),
+	DAILINK_COMP_ARRAY(COMP_DUMMY()),
+	DAILINK_COMP_ARRAY(COMP_EMPTY()));
+
+SND_SOC_DAILINK_DEFS(be_iis1,
+	DAILINK_COMP_ARRAY(COMP_CPU("VBC_IIS1")),
+	DAILINK_COMP_ARRAY(COMP_DUMMY()),
+	DAILINK_COMP_ARRAY(COMP_EMPTY()));
+
+SND_SOC_DAILINK_DEFS(be_iis2,
+	DAILINK_COMP_ARRAY(COMP_CPU("VBC_IIS2")),
+	DAILINK_COMP_ARRAY(COMP_DUMMY()),
+	DAILINK_COMP_ARRAY(COMP_EMPTY()));
+
+SND_SOC_DAILINK_DEFS(be_iis3,
+	DAILINK_COMP_ARRAY(COMP_CPU("VBC_IIS3")),
+	DAILINK_COMP_ARRAY(COMP_DUMMY()),
+	DAILINK_COMP_ARRAY(COMP_EMPTY()));
+
+static const struct snd_soc_pcm_stream voice_params = {
+	.formats = SNDRV_PCM_FMTBIT_S16_LE,
+	.rate_min = 8000,
+	.rate_max = 8000,
+	.channels_min = 1,
+	.channels_max = 1,
+};
+
+static struct snd_soc_dai_link ums9230_card_dai_links[] = {
+	/* DPCM frontend */
+	{
+		.name = "FE_FAST_PLAYBACK",
+		.stream_name = "FE_FAST_PLAYBACK",
+		.trigger = {SND_SOC_DPCM_TRIGGER_PRE,
+			    SND_SOC_DPCM_TRIGGER_PRE},
+		.dynamic = 1,
+		.playback_only = 1,
+		SND_SOC_DAILINK_REG(fe_fast_playback),
+	},
+	{
+		.name = "FE_CAPTURE_DSP",
+		.stream_name = "FE_CAPTURE_DSP",
+		.trigger = {SND_SOC_DPCM_TRIGGER_PRE,
+			    SND_SOC_DPCM_TRIGGER_PRE},
+		.dynamic = 1,
+		.capture_only = 1,
+		SND_SOC_DAILINK_REG(fe_capture_dsp),
+	},
+
+	/* Modem voice channels */
+	{
+		.name = "VOICE",
+		.stream_name = "VOICE",
+		.c2c_params = &voice_params,
+		.num_c2c_params = 1,
+		SND_SOC_DAILINK_REG(voice_hostless),
+	},
+	{
+		.name = "VOICE_CAPTURE",
+		.stream_name = "VOICE_CAPTURE",
+		.capture_only = 1,
+		SND_SOC_DAILINK_REG(voice_capture),
+	},
+
+	/* DPCM backend */
+	{
+		.name = "IIS0",
+		.dai_fmt = SND_SOC_DAIFMT_I2S | SND_SOC_DAIFMT_NB_NF |
+			   SND_SOC_DAIFMT_CBP_CFP,
+		.no_pcm = 1,
+		.ignore_suspend = 1,
+		SND_SOC_DAILINK_REG(be_iis0),
+	},
+	{
+		.name = "IIS1",
+		.dai_fmt = SND_SOC_DAIFMT_I2S | SND_SOC_DAIFMT_NB_NF |
+			   SND_SOC_DAIFMT_CBP_CFP,
+		.no_pcm = 1,
+		.ignore_suspend = 1,
+		SND_SOC_DAILINK_REG(be_iis1),
+	},
+	{
+		.name = "IIS2",
+		.dai_fmt = SND_SOC_DAIFMT_I2S | SND_SOC_DAIFMT_NB_NF |
+			   SND_SOC_DAIFMT_CBP_CFP,
+		.no_pcm = 1,
+		.ignore_suspend = 1,
+		SND_SOC_DAILINK_REG(be_iis2),
+	},
+	{
+		.name = "IIS3",
+		.dai_fmt = SND_SOC_DAIFMT_I2S | SND_SOC_DAIFMT_NB_NF |
+			   SND_SOC_DAIFMT_CBP_CFP,
+		.no_pcm = 1,
+		.ignore_suspend = 1,
+		SND_SOC_DAILINK_REG(be_iis3),
+	},
+};
+
+static struct snd_soc_card ums9230_card = {
+	.dai_link = ums9230_card_dai_links,
+	.num_links = ARRAY_SIZE(ums9230_card_dai_links),
+	.owner = THIS_MODULE,
+};
+
+static int ums9230_card_probe(struct platform_device *pdev)
+{
+	struct device *dev = &pdev->dev;
+	struct snd_soc_card *card;
+	int ret;
+
+	card = &ums9230_card;
+	card->dev = dev;
+
+	ret = sprd_card_parse_of(card);
+	if (ret)
+		return ret;
+
+	return devm_snd_soc_register_card(dev, card);
+}
+
+static const struct of_device_id ums9230_card_of_match[] = {
+	{ .compatible = "sprd,ums9230-audio-card" },
+	{ }
+};
+MODULE_DEVICE_TABLE(of, ums9230_card_of_match);
+
+static struct platform_driver ums9230_card_driver = {
+	.driver = {
+		.name = "snd-soc-ums9230-card",
+		.of_match_table = ums9230_card_of_match,
+	},
+	.probe = ums9230_card_probe,
+};
+
+module_platform_driver(ums9230_card_driver);
+
+MODULE_DESCRIPTION("Unisoc ASoC Machine Driver");
+MODULE_LICENSE("GPL");
